ETK Help Center: Create a tabs panel for next steps tutorials

jeyip opened the pull request:
#### Proposed Changes

* Add a tabs panel to the help center to prepare a new section for next steps tutorials

#### Mocks
 
<img width=316 src="https://user-images.githubusercontent.com/195089/171740786-c5e20605-9fec-45fa-94db-dab93d1c0b45.png" />

#### GIFs
![2022-06-24 11 48 08](https://user-images.githubusercontent.com/5414230/175646842-a65a9e39-75b7-46b2-b223-ef19ed9e90c2.gif)

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Sandbox a site
* Navigate to `apps/editing-toolkit`
* You may need to run `yarn`
* Run `yarn dev --sync` 
* Navigate to sandboxed site in the browser
* Open the help center menu from the top right hand corner of the editor ( it's a question mark icon )

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/64260/ and https://github.com/Automattic/wp-calypso/issues/64080

jeyip committed:
Create a tab panel for next steps

jeyip committed:
Style tab panel component

jeyip commented on packages/help-center/src/styles.scss:
```diff
@@ -62,6 +63,17 @@ $head-foot-height: 50px;
 		}
 	}
 
+	.help-center-header__tab-panel {
```
**Note:**

The designs from the mocks diverge from the tab styling of the shared component. I override core tab styling here.

jeyip commented on packages/help-center/src/components/help-center-header.tsx:
```diff
@@ -70,56 +71,87 @@ const HelpCenterHeader: React.FC< Header > = ( {
+					} }
+				>
+					{ () => {
+						return <></>;
```
The TabPanel is being super fussy 😞

It demands a direct child prop, or else it throws a runtime error, which is why I pass this no-op here. The reason we don't need direct children is because the help center feature already implements a `react-router` MemoryRouter to dynamically render content.

I'm thinking about adding a condition to the shared component to handle this, but curious if anyone else has thoughts on this  🤷

jeyip replied:
> It demands a direct child prop, or else it throws a runtime error, which is why I pass this no-op here. The reason we don't need direct children is because the help center feature already implements a react-router MemoryRouter to dynamically render content.

To elaborate on why I opted to not use direct children props, I'll describe the alternative approach where we do use them.

```js
// help-center-container.tsx

// existing implementation in this PR
<Card>
  <HelpCenterHeader /> // <--- new TabPanel component lives in here
  <HelpCenterContent /> // <--- NextStepsTutorials view lives in here
  <HelpCenterFooter />
</Card>

// alternative implementation
<Card>
  <HelpCenterHeader />
  <TabPanel>
    { ( tab ) => {
      if ( tab === 'nextSteps' ) {
        return <NextStepsTutorials />
      } else if ( tab === 'help' ) {
        return (
          <>
            <HelpCenterContent /> 
            <HelpCenterFooter />
          </>
        )
      }
    } }
  <TabPanel />
</Card>
```
- `HelpCenterContent` already has a `Switch` and `Routes`, and its entire purpose is to use `react-router` to alternate between all possible views that can be displayed in the help center ( contact form, contact options, inline chat, etc. )
- In this PR, we simply add one more route to `HelpCenterContent` for the Next Steps Tutorial view.
- If we, instead, pass children directly to the tab panel, we're agreeing to mix state management of the views between `TabPanel` and `react-router`

Addison-Stavlo replied:
Thats for detailing that. I think what you are doing makes sense with the current routing setup.

Im still a little unsure why we are routing at all here (previous to this PR) when it seems like everything could fit into the tab-group/tab-panel/tab-children component paradigm and piggyback off core for things like a11y interactions between those levels. Is the routing buying us something I am missing?

jeyip replied:
> Im still a little unsure why we are routing at all here (previous to this PR) when it seems like everything could fit into the tab-group/tab-panel/tab-children component paradigm

I might be missing something, but the routing actually makes sense to me. Tabs are traditionally meant to switch between content at identical hierarchical levels. Some of the views in the help center reflect a nested navigation structure, which, to me, lends itself to a solution like `react-router`

Ex. We navigate to the inline chat view by clicking on the "Still need help?" button _and then_ the "Live Chat" button
![2022-06-27 10 22 04](https://user-images.githubusercontent.com/5414230/175999204-bacbbef4-3a44-4d82-af8e-c40be68bbfaf.gif)

I didn't see elaboration on the reasoning in the original PR that migrated logic to `react-router` though, so @alshakero might be able to speak more to this discussion

Addison-Stavlo replied:
I don't think its necessarily an issue but was a bit curious.

> Tabs are traditionally meant to switch between content at identical hierarchical levels. Some of the views in the help center reflect a nested navigation structure, which, to me, lends itself to a solution like react-router

At my initial look, I would see the tabs and their content containers as entirely separated concerns from one another, and see the tab-content containers at identical hierarchical levels WRT the tab interface regardless of how their internal content is managed.

 The tab for 'Help' has tab content that has some nested views levels and would require some sort of state/view management (component level, redux, router, etc.). Then there is the tab for "Next Steps" which has its own content which may require its own internal levels of navigation. 

But neither of these tab content containers share any of the same levels and views (or did i miss this somewhere maybe?).  We are never going to be on the 'Help' tab and needing to show content from 'Next Steps' of vice-versa. Their contents are entirely separate and independent. So my initial expectations would be to see the state management solutions for these contents internal to the tab content containers they are concerned with, as opposed to wrapping the entire system with a router and controlling the tabs themselves with it.

Addison-Stavlo replied:
All that said, thats just my initial thoughts on how I would go about managing the tabs and their content. I don't think the router this way is necessarily a bad approach or a problem, as long as it works and we are confident development on either side wont have adverse effects on the other.

jeyip replied:
Ah I forgot to reply to this yesterday 🤦 Sorry Addie! I'll try to be more thorough moving forward.

> So my initial expectations would be to see the state management solutions for these contents internal to the tab content containers they are concerned with, as opposed to wrapping the entire system with a router and controlling the tabs themselves with it.

I think this is a fair point. Giving it more thought, I don't feel very strongly about either the react-router or the tabs panel local state solution, because, if, as you say, "We are never going to be on the 'Help' tab and needing to show content from 'Next Steps' of vice-versa," then it will be easy to refactor one solution into the other in the future.

I'm fine switching over to the alternative implementation as described in https://github.com/Automattic/wp-calypso/pull/64879#discussion_r906437778 if we'd prefer that. We'll probably want to rename files in that case, which we can do in a follow-up.

@escapemanuele thoughts?

mattwiebe replied:
My main thought here is flows that re-enter the Help Center after it's been closed. Will state be maintained? Is that desirable?

My main scenario is like this:

1. The Help Center closes when a Tour is started. 
2. The Tour is completed and Help Center opens to cross off a Step. 
3. Is the Next Steps tab maintained or does it reset to Help? Currently it resets when re-opened.

If routing helps us to maintain that state, we should use it. I'm sure we can maintain that state without it, too. And we should probably not maintain state when we _haven't_ started a Tour after clicking into Next Steps. Tricky.

Addison-Stavlo replied:
> Ah I forgot to reply to this yesterday 🤦 Sorry Addie! I'll try to be more thorough moving forward.

No worries at all, I see this thread as more of a discussion and not any urgent blocker.

> I'm fine switching over to the alternative implementation as described in https://github.com/Automattic/wp-calypso/pull/64879#discussion_r906437778 if we'd prefer that.

What do you think makes sense? Im fine keeping it the way it is now if we think it makes sense and will continue to work well for our uses. As you noted if we do find need to change it later on

> then it will be easy to refactor one solution into the other in the future.

Addison-Stavlo replied:
>  is flows that re-enter the Help Center after it's been closed. Will state be maintained? Is that desirable?

Good questions and important to consider! I think both approaches give us ways to handle this, but perhaps the router makes it easier to manage.

We can leave this change to a follow-up issue, and will probably need some confirmation from design. I opened a P2 and pinged @nuriapenya and @Automattic/vertex  team for thoughts. paYE8P-1Ii-p2

jeyip replied:
> What do you think makes sense?

I'm in camp `react-router`

I agree that local state management makes the most intuitive sense as a developer, but on the flip side, I don't see the usage of `react-router` for these tab panels as unintuitive.

> We are never going to be on the 'Help' tab and needing to show content from 'Next Steps' of vice-versa. Their contents are entirely separate and independent. 

Although not currently designed or on a roadmap, I also wouldn't be surprised if views _do_ eventually become shared within each of the tabs as the help center evolves.

Ex. We might add a "Get Help" button to the next steps tutorial panel that opens the live chat window ( which "lives" in the help tab ). In this case, state managment with `react-router` would make the issue trivial -- create a link component that switches routes. If we were using local state management, we'd need to know to programatically render the help tab first to ensure the inline chat view is available _and then_  switch routes.

Also, as mentioned above, if we end up needing to maintain Help Center state after it's been closed, `react-router` will probably make that easier as well.

nuriapenya replied:
> Is the Next Steps tab maintained or does it reset to Help? Currently it resets when re-opened.

I added [some thoughts](https://cylonp2.wordpress.com/2022/06/30/help-center-maintaining-state-on-re-open/#comment-1577) on the post but essentially, If we can maintain the last state of Next Steps when reopening the modal, that would be great for the users - so they can pick up on the last thing they did.

> And we should probably not maintain state when we haven't started a Tour after clicking into Next Steps. 

I also agree with that. If the user did not engage at all with Next Steps, perhaps we can just reset it back to Help Center. If that make sense?

Addison-Stavlo replied:
Lets leave the state preservation to a follow-up PR. (created [an issue here](https://github.com/Automattic/wp-calypso/issues/65160))

>  If the user did not engage at all with Next Steps, perhaps we can just reset it back to Help Center. If that make sense?

@nuriapenya - It does, but I wonder if its that important? From a technical standpoint, updating it to maintain its state between closing/opening is a rather simple concern - essentially keeping the components 'aware' even when they aren't visible. Something we will also need in the future if we want to provide links from MyHome or LaunchPad to send users into the editor and kickstart specific tutorial flows.  

However, trying to track user actions and evaluating what they did and did not engage with since the last time they selected a tab to determine what tab we should route them to the next time the center is opened seems like it will be a bit of a complicated system open to some awkward edge cases and weird bugs - for not very much benefit. If we keep it simple and a user clicks "next steps" tab, doesn't do anything within that tab, closes and reopens the help center in the same editing session, and is still on the "next steps" tab... I don't think thats much of an issue. 

Mind you this state preservation im speaking of is within the course of 1 editor session, not after closing and re-opening the editor. When that is the case I think it makes sense to have "Help" as the default landing state, unless we are talking about a specific entry flow that is intended to kick-off a tutorial.

nuriapenya replied:
>If we keep it simple and a user clicks "next steps" tab, doesn't do anything within that tab, closes and reopens the help center in the same editing session, and is still on the "next steps" tab... I don't think thats much of an issue.

True. I thought resetting back to the Help centre was the simple approach.

> Mind you this state preservation im speaking of is within the course of 1 editor session, not after closing and re-opening the editor. 

Gotcha 👌🏽 That makes total sense, it should be always while users are interacting with the Editor.

jeyip commented:
Currently fixing a bug where the heading text disappears when the help center is minimized 🤔 

Everything else is ready for a first look though

Addison-Stavlo reviewed (commented):
Im curious on your thoughts regarding how we should develop this feature/release this feature. Obviously we don't want to merge the new tab panel in before we have tours ready. 
Do you think we should maintain a feature branch? Or maybe we can we add in a proxied check for our "Next Steps" sections until we are ready to release the tours? Or maybe you have some other/better ideas? 🤔 

------------------

Looking at **a11y at the keyboard navigation level**:
* The tab-buttons have no visual indicator for focus. When the panel opens, focus goes into the search box of the 'help' tab. We can get back to the tab-button-level with `shft+tab` and between them with L and R arrowkeys. Although, its hard to tell that is the case since these tab-buttons seem to be the only points of this interface without some focus style.
* This may not be an issue, but switching from the 'Next Steps' tab to the 'Help' tab automatically moves focus into the content container (search input) instead of keeping focus at the tab-button level until the user is ready to engage with the content of that tab group. With only 2 tab-buttons this seems possibly acceptable (?), but if we ever get into having a 3rd tab this could be quite frustrating for keyboard navigation. Thoughts?

Addison-Stavlo reviewed (commented):
> Currently fixing a bug where the heading text disappears when the help center is minimized

I didnt notice this myself, maybe I am mis-interpereting?

I did however find an odd bug around the minimize/re-open flow. If the "Next Steps" is selected when minimized, reopening the container results in a mismatch between highlighted tabs and the shown content (shows next steps content but highlights 'help' tab):
![opens-wront-tab-highlight](https://user-images.githubusercontent.com/28742426/175941384-7d8aef13-e717-40fe-b10a-f579d29be5dd.gif)

jeyip committed:
Account for react router in initial tab state

React router holds in memory state of which view was last displayed in the help center. When we minimize the help center, react router will re-render the last view we saw, rather than resetting things. The tab panel, however, is unmounted -- selected tab state is reset.

To avoid mismatches between the selected tab and which view is being displayed, we now take into account the current react router pathname to set initial selected tab state

jeyip commented:
Sorry everyone! Borked the rebase 😅

jeyip commented:
> I did however find an odd bug around the minimize/re-open flow. If the "Next Steps" is selected when minimized, reopening the container results in a mismatch between highlighted tabs and the shown content (shows next steps content but highlights 'help' tab):

Should be fixed in [d114f49](https://github.com/Automattic/wp-calypso/pull/64879/commits/d114f49a2ff557bf6d5293a593cb0388d99d6992)

jeyip committed:
Add help center header when next steps view is minimized

jeyip commented:
> Currently fixing a bug where the heading text disappears when the help center is minimized 🤔

Fixed in [899a6f6](https://github.com/Automattic/wp-calypso/pull/64879/commits/899a6f60f0d9210b9853e58db46cd2265f3af69a)

jeyip committed:
Fix outlines for tab panels

jeyip commented:
> The tab-buttons have no visual indicator for focus. When the panel opens, focus goes into the search box of the 'help' tab. We can get back to the tab-button-level with shft+tab and between them with L and R arrowkeys. Although, its hard to tell that is the case since these tab-buttons seem to be the only points of this interface without some focus style.

Fixed in [adb0daa](https://github.com/Automattic/wp-calypso/pull/64879/commits/adb0daaf7c89fb4f40e2029b523f11b4e2696d3e)

jeyip commented:
> This may not be an issue, but switching from the 'Next Steps' tab to the 'Help' tab automatically moves focus into the content container (search input) instead of keeping focus at the tab-button level until the user is ready to engage with the content of that tab group. With only 2 tab-buttons this seems possibly acceptable (?), but if we ever get into having a 3rd tab this could be quite frustrating for keyboard navigation. Thoughts?

I'm not sure if I inadvertently fixed this, but I can't seem to reproduce the behavior. I tried using arrow keys to switch between tabs, and the focus seems to stay on the tab panels. I also tried using the mouse to switch as well, which appears to be fine. 🤔

jeyip committed:
Use wp admin theme color for tab styling

jeyip commented:
>  think the other thing to note here (that Nuria noted in slack thread last night) is the colors we use for borders/buttons/etc. (red in the mock up) should reflect the site's default color scheme (like how the save button, block inserter button, and other interface elements do). Right now where my test site's color scheme is blue, I see everything as blue except for the active-tab indicator style.

Fixed in [6937bb9](https://github.com/Automattic/wp-calypso/pull/64879/commits/6937bb9edc4a5a8dba90e87323cb3b0341c8c936)

jeyip commented:
> Are we down to the last remaining concern being the release workflow stuff for this PR?

I'm seeing a variety of odd CI build errors, but they don't appear to be related to this PR. There's also final question of deciding where we want to handle state management of the tabs ( locally vs react-router ), but I'm not super concerned about this since either direction will work and things are easily reversible. Going to respond in more detail in the comment thread in a moment.

Besides that I agree that we're in a good place and can merge this when we have workflow stuff sorted out.

jeyip committed:
Add feature toggle to next steps tabs

jeyip commented:
> Are we down to the last remaining concern being the release workflow stuff for this PR?

Added a feature toggle in https://github.com/Automattic/wp-calypso/pull/64879/commits/bfdd2627705541de2998f5e3345a0c0e574e15b8

jeyip committed:
Undo whitespace change

jeyip committed:
Fix typescript error

jeyip commented:
Deprioritizing and closing this for now as we shift our focus to the launchpad feature https://createp2.wordpress.com/2022/07/06/create-roadmap-update-2/
